File: java/src/main/java/vsdk/toolkit/render/RenderHookList.java
package vsdk.toolkit.render;

import java.util.ArrayList;

/**
Render hooks are called each time the scene is rendered.
Functions are provided to add and remove hooks.
Hooks should only depend on render.h, not on GLX or OpenGL
*/
public class RenderHookList {
    private static ArrayList<RenderHook> renderHookList = new ArrayList<>();

    public static void renderHooks() {
        for (int i = 0; renderHookList != null && i < renderHookList.size(); i++) {
            RenderHook h = renderHookList.get(i);
            h.function.apply(h.data);
        }
    }

    public static void removeAllRenderHooks() {
        renderHookList = null;
    }
}

File: turboc/src/render/RenderHookList.cpp
#include "java/util/ArrayList.txx"
#include "render/RenderHookList.h"
#include "render/RenderHook.h"

ArrayList<RenderHook *> *RenderHookList::renderHookList = new ArrayList<RenderHook*>();

void
RenderHookList::renderHooks() {
    for ( int i = 0; renderHookList != NULL && i < renderHookList->size(); i++ ) {
        RenderHook *h = renderHookList->get(i);
        h->function(h->data);
    }
}

void
RenderHookList::removeAllRenderHooks() {
    delete renderHookList;
    renderHookList = NULL;
}
